routers/web/user/code.go

```
@@ -0,0 +1,111 @@
+		return
+	}
+
+	ctx.Data["IsPackageEnabled"] = setting.Packages.Enabled
+	ctx.Data["IsRepoIndexerEnabled"] = setting.Indexer.RepoIndexerEnabled
+	ctx.Data["Title"] = ctx.Tr("code.title")
+	ctx.Data["ContextUser"] = ctx.ContextUser
+
+	language := ctx.FormTrim("l")
+	keyword := ctx.FormTrim("q")
+	page := ctx.FormInt("page")
+	if page <= 0 {
+		page = 1
+	}
+
+	queryType := ctx.FormTrim("t")
+	isMatch := queryType == "match"
+
+	opts := &repo_model.SearchRepoOptions{
+		Collaborate: util.OptionalBoolFalse,
+		Actor:       ctx.Doer,
+		OwnerID:     ctx.ContextUser.ID,
+	}
+	if ctx.IsSigned {
+		opts.Private = true
+	}
+	repoIDs, _, err := repo_model.SearchRepositoryIDs(opts)
+	if err != nil {
+		ctx.ServerError("SearchResults", err)
+		return
+	}
+
+	repoMaps, err := repo_model.GetRepositoriesMapByIDs(repoIDs)
+	if err != nil {
+		ctx.ServerError("SearchResults", err)
+		return
+	}
+
+	// if non-admin login user, we need check UnitTypeCode at first
+	if ctx.Doer != nil && !ctx.Doer.IsAdmin && len(repoIDs) > 0 {
+		rightRepoMap := make(map[int64]*repo_model.Repository, len(repoMaps))
+		repoIDs = make([]int64, 0, len(repoMaps))
+		for id, repo := range repoMaps {
+			if models.CheckRepoUnitUser(ctx, repo, ctx.Doer, unit.TypeCode) {
+				rightRepoMap[id] = repo
+				repoIDs = append(repoIDs, id)
+			}
+		}
+		ctx.Data["RepoMaps"] = rightRepoMap
+	} else {
+		ctx.Data["RepoMaps"] = repoMaps
+	}
+
+	var (
+		total                 int
+		searchResults         []*code_indexer.Result
+		searchResultLanguages []*code_indexer.SearchResultLanguages
+	)
+
+	total, searchResults, searchResultLanguages, err = code_indexer.PerformSearch(ctx, repoIDs, language, keyword, page, setting.UI.RepoSearchPagingNum, isMatch)
+	if err != nil {
+		if code_indexer.IsAvailable() {
+			ctx.ServerError("SearchResults", err)
+			return
+		}
+		ctx.Data["CodeIndexerUnavailable"] = true
+	} else {
+		ctx.Data["CodeIndexerUnavailable"] = !code_indexer.IsAvailable()
+	}
```

I know this part is only copy pasted but shouldn't `code_indexer.IsAvailable()` get called once at the beginning to fail fast if you can't use the code search?

---

no, it will try to search (elastic could be available already) only when it returns error, we check if that is because elastic is not available